StreamChatClient_v3/Database/DTOs/MessageDTO.swift

```diff
@@ -7,14 +7,103 @@ import CoreData
 import Foundation
 
 @objc(MessageDTO)
-public class MessageDTO: NSManagedObject {
+class MessageDTO: NSManagedObject {
     static let entityName = "MessageDTO"
     
-    @NSManaged fileprivate var id: String
-    @NSManaged fileprivate var text: String
-    @NSManaged fileprivate var additionalStateRaw: Int16
-    @NSManaged fileprivate var timestamp: Date
+    @NSManaged var additionalStateRaw: Int16
+    @NSManaged var id: String
+    @NSManaged var text: String
+    @NSManaged var type: String
+    @NSManaged var command: String?
+    @NSManaged var createdDate: Date
+    @NSManaged var updatedDate: Date
+    @NSManaged var deletedDate: Date?
+    @NSManaged var args: String?
+    @NSManaged var parentId: String?
+    @NSManaged var showReplyInChannel: Bool
+    @NSManaged var replyCount: Int32
+    @NSManaged var extraData: Data
+    @NSManaged var isSilent: Bool
+    @NSManaged var reactionScores: NSDictionary
     
-    @NSManaged fileprivate var user: UserDTO
-    @NSManaged fileprivate var channelId: String?
+    @NSManaged var user: UserDTO
+    @NSManaged var mentionedUsers: Set<UserDTO>
+    @NSManaged var channel: ChannelDTO
+    
+    static func load(for cid: String, limit: Int, context: NSManagedObjectContext) -> [MessageDTO] {
+        let request = NSFetchRequest<MessageDTO>(entityName: entityName)
+        request.predicate = NSPredicate(format: "channel.cid == %@", cid)
+        request.sortDescriptors = [NSSortDescriptor(keyPath: \MessageDTO.updatedDate, ascending: false)]
+        request.fetchLimit = limit
+        return (try? context.fetch(request)) ?? []
+    }
+    
+    static func load(id: String, context: NSManagedObjectContext) -> MessageDTO? {
+        let request = NSFetchRequest<MessageDTO>(entityName: entityName)
+        request.predicate = NSPredicate(format: "id == %@", id)
+        return try? context.fetch(request).first
+    }
+    
+    static func loadOrCreate(id: String, context: NSManagedObjectContext) -> MessageDTO {
+        if let existing = load(id: id, context: context) {
+            return existing
+        }
+        
+        let new = NSEntityDescription.insertNewObject(forEntityName: entityName, into: context) as! Self
+        new.id = id
+        return new
+    }
```

Comment:
I'm actually thinking about extending `NSManagedObject` and adding these methods, since they'll be awfully similar for each object

Reply:
They differ in the predicate and also in what field exactly do we treat as the unique id. You can definitely try but I think it will be a little bit of hell to make this work with protocols :)

Reply:
Otherwise, I would have already done that :D :D